Add client steering apis

itayx opened the pull request:
The client steering APIs will help handle and manage the clients in the soon to be persistent DB.

The client functions allow the user to do the following:
- Receive a list of available clients using the `get_client_list` function.
- Update an existing client using their station's MAC address using the "set_client" function.
- Retrieve an existing client using their station's MAC address using the "get_client" function.

itayx committed:
bcl: add BML structs const defines

preparative commit

As part of this PR we introduce persistent data with the ability to be
set to "not configured" meaning the data is not stored in the persistent
database. Both will be used later by sClientConfig & sClient.

Add define for PARAMETER_NOT_CONFIGURED.
Add enum-like defines for the client's Selected-Bands
Add structures for the client configuration.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
tlvf: add BML_CLIENT messages for smart-steering

preparative commit

Add client set/get CMDUs to enable smart-steering

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
son_management: add client smart-steering support

To support the BML calls, the son_management is responsible for the
messages handling back and forth from the BML.

For the mean time the son_management will only handle the messages with
no internal logic.

Add support for the smart-steering APIs in son_management.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -492,3 +492,48 @@ cACTION_BML_WIFI_CREDENTIALS_CLEAR_REQUEST:
+  result:
+    _type: uint8_t
+    _comment: # 0 - Failure, 1 - Success
+  
```
remove multiple line spacing

itayx replied:
removed

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,52 @@ sBssidInfo:
+    _comment: 1 for true, 0 for false, -1 for "not configured".
+  time_life_delay_days:
+    _type: uint32_t
+    _comment: optional parameter, Timelife delay of the client configuration (if configured), otherwise 0.
```
need to add empty new line at end of file

itayx replied:
Added

adam1985d commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,61 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
+
+    auto pBML = static_cast<bml_internal *>(ctx);
+    return pBML->client_dump_client(tlvf::mac_from_string(std::string(sta_mac)));
+}
```
need to add empty new line at end of file

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,61 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
+int bml_client_get_client_list(BML_CTX ctx, unsigned char *client_list,
+                               unsigned int *client_list_size)
+{
+    // Validate intput params;
```
I don't understand the need for `;` at end of comment.
I know you probably copied it from the DCS APIs, but the DCS APIs have 2 kinds of comments:
`// Validate input parameters`
and
`// Validate intput params;` (the mistake is in the origin)

either way, remove the `;` and fix all locations of `intput` to `input`

itayx replied:
Yeah, copy paste mistake

adam1985d commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,61 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
                                        channel_pool, channel_pool_size);
 }
+
+int bml_client_get_client_list(BML_CTX ctx, unsigned char *client_list,
```
we discussed that the mac-list will be of signed-char since this is what we see in other radio-mac and al-mac function parameters - please align the APIs to that (and later adjust the HLD accordingly)
(other BML APIs are already aligned in the HLD and in your PR)

itayx replied:
Aligned

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    int stay_on_selected_device;
+    //comma-separated selected bands: 2.4,5H,5L,6.
+    char selected_bands[BML_SELECTED_BANDS_LEN];
+    //1 for true, 0 for false, -1 for "parameter is not configured".
```
add space after `//`
(verify on all the comments)

itayx replied:
Fixed

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1127,6 +1127,81 @@ int bml_internal::process_cmdu_header(std::shared_ptr<beerocks_header> beerocks_
+                    m_client->timestamp_sec           = response->client().timestamp_sec;
+                    m_client->stay_on_initial_radio   = response->client().stay_on_initial_radio;
+                    m_client->stay_on_selected_device = response->client().stay_on_selected_device;
+                    mapf::utils::copy_string(response->client().selected_bands,
```
I am not sure if in the CMDU we want to pass a string or a bitwise of selected bands and translate it

itayx replied:
@kantera800 @morantr WTYT?

morantr replied:
I Always prefer using enums over strings if possible, bitwise are ok too if you do it right.
If you need something like this, bitwise is a good idea:

band_5G | band_2.4G | output
-- | -- | --
0 | 0 | All bands disabled
0 | 1 | Only band 2.4G
1 | 0 | Only band 5G
1 | 1 | Both bands 2.4G and 5G

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,52 @@ sBssidInfo:
+    _type: uint32_t
+    _comment: Time of last client update (in Seconds)
+  stay_on_initial_radio:
+    _type: uint8_t
```
replace `uint8_t` with `int8_t` since this parameter can receive `-1`
(this comment is related to all tlvf parameters that support `-1` as "not-configured" value

itayx replied:
You're right, my mistake, changed

adam1985d commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,51 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * Get client list.
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [out] client_list         Client list.
```
this is actually an `[in/out]` parameter (the calling user need to provide an already allocated array with the max size provided by the `client_list_size` which is also `[in/out]`
(see bml_get_vap_list_credentials as an example)

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,51 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [in] sta_mac              MAC address of station.
+ * @param [out] client              Client information.
```
the client struct should be allocated by calling user and parameter should be `[in/out]`

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -3282,4 +3527,4 @@ bool bml_internal::wake_up(uint8_t action_opcode, int value)
     m_prmCliResponses[action_opcode]->set_value(value);
     return true;
-}
+}
```
need to add empty new line at end of file

itayx replied:
Added

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1127,6 +1127,81 @@ int bml_internal::process_cmdu_header(std::shared_ptr<beerocks_header> beerocks_
+            if (m_prmClientListGet) {
+                if (m_client_list && m_client_list_size) {
+                    *m_client_list_size = response->client_list_size();
+                    *m_client_list      = response->client_list_str();
```
do we need here to copy the string or just point to it with the pointer as you did?

itayx replied:
I didn't
```
int N = 5;
auto P = &N; //Pointer to N
*P = 6; //Will change the value of the pointer, not the pointer itself
```
So `*m_client_list` will use the value of `m_client_list` (which is a string type) and run the `string::operator=` which copies the data.

adam1985d replied:
ok

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -480,4 +480,13 @@ eActionOp_BML:
   ACTION_BML_CHANNEL_SCAN_DUMP_RESULTS_RESPONSE: 213
 
-  ACTION_BML_ENUM_END: 214
+  ACTION_BML_GET_CLIENT_LIST_REQUEST: 215
```
I think that just like we changed the `bml_set_client..` to be `bml_client_set_client..` for grouping the APIs,
we can also change the CMDUs from `ACTION_BML_GET_CLIENT...` to `ACTION_BML_CLIENT_GET_CLIENT...`

nitpick

itayx replied:
Changed, this makes the CMDUs more understandable

adam1985d reviewed (changes requested):
the PR is good, but I gave several comments and want to re-review them - so setting "request changes"

itayx committed:
controller: bml: add client smart-steering APIs

This PR enables the smart-steering feature which allows a user to
configure a client's steering behavior.
The configuration is persistent across prplmesh restarts and reboots.

Added BML API to get list of configured clients.
Added BML client set and get APIs.
Added BML-internal implementation

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    int stay_on_initial_radio;
+    // 1 for true, 0 for false, -1 for "not configured".
+    int stay_on_selected_device;
+    // comma-separated selected bands: 2.4,5H,5L,60.
```
I misled you after I misled myself,
the value should be 6 and not 60. please revert back the change

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    int stay_on_initial_radio;
+    // 1 for true, 0 for false, -1 for "not configured".
+    int stay_on_selected_device;
+    // comma-separated selected bands: 2.4,5H,5L,60.
```
I misled you after I misled myself,
the value should be 6 and not 60. please revert back the change

kantera800 replied:
actually both bands exists but i think  that 6Ghz is the one that we need to consider 
**80211.ad:** around 60 GHz band (58.32 - 69.12 GHz)   
**80211.ax:** around 6 GHz band (5.9 - 7.2 GHz)

we could just add both and consider them as not supported

arnout replied:
Are we going to consider 6GHz as a separate band? (Note: I mean in general - here it does make sense to consider 6 separate, just like 5L and 5H.)

Currently, we mostly assume that a radio operates on only a single band. However, AFAIU, an 11ax radio will typically operate on 5 and 6GHz simultaneously. So it doesn't make a lot of sense to see it as a separate band from 5GHz.

I like more and more how EasyMesh works only with operating classes, not with bands or channels or bandwidth or anything like that :-)

simonbaren commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,95 @@ void son_management::handle_bml_message(Socket *sd,
+            message_com::create_vs_message<beerocks_message::cACTION_BML_GET_CLIENT_LIST_RESPONSE>(
+                cmdu_tx, beerocks_header->id());
+        if (!response) {
+            LOG(ERROR) << "Failed building message "
```
possible segfault - you should break from here. you later dereference `response` at line 2332

itayx replied:
Added break

simonbaren commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,95 @@ void son_management::handle_bml_message(Socket *sd,
+
+        //To make sure the response is handled using a faux empty list.
+        auto faux_client_list      = "";
+        auto faux_client_list_size = sizeof(faux_client_list);
```
'faux_client_list' is probably a std::string. If I understand what you are trying to do here - you should use `faux_client_list.size()` instead.

itayx replied:
Changed

simonbaren commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,95 @@ void son_management::handle_bml_message(Socket *sd,
+                cmdu_tx, beerocks_header->id());
+        if (!response) {
+            LOG(ERROR) << "Failed building message "
+                          "cACTION_BML_SET_CLIENT_RESPONSE !";
```
break?

itayx replied:
Added break

simonbaren commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,52 @@ sBssidInfo:
+  _type: struct
+  stay_on_initial_radio:
+    _type: int8_t
+    _comment: 1 for true, 0 for false, -1 for "not configured".
```
Looks like you use the 1,0,-1 a lot, I'd suggest an enum for this.

itayx replied:
Tried using an enum, but the yaml doesn't accept an enum with a `-1` value

itayx committed:
beerocks_cli: add smart-steering client APIs.

In this PR we introduced new BML APIs for the smart-steering and
persistent DB feature.

Add CLI calls to the new smart-steering APIs:
bml_client_get_client_list - Retrieve full client list.
bml_client_set_client - Update client by STA MAC.
bml_client_get_client - Retrieve client by STA MAC.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

simonbaren commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,98 @@ void son_management::handle_bml_message(Socket *sd,
+        //To make sure the response is handled using a faux empty list.
+        std::string faux_client_list = "";
+        if (!response->alloc_client_list(faux_client_list.size())) {
+            LOG(ERROR) << "Failed client_list allocation to size=" << int(faux_client_list.size());
```
break

itayx replied:
Fixed

morantr commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -70,6 +70,7 @@ enum eStructsConsts {
     HT_MCS_SET_SIZE                      = 16,
     VHT_MCS_SET_SIZE                     = 32,
     DEV_MAX_RADIOS                       = 3,
+    SELECTED_BANDS_LENGTH                = 32,
```
It is probably best to not define it and use it as a string.  You could use enum or bitfield.

itayx replied:
Changed to use enum

morantr commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
 
+struct BML_CLIENT_CONFIG {
+    // 1 for true, 0 for false, -1 for "not configured".
+    int stay_on_initial_radio;
```
You used `int` to all of the tri-state parameters. Please use `int8_t`. It cost 4 times less memory. :-)

itayx replied:
Fixed

morantr commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,52 @@ sBssidInfo:
+    _comment: Client MAC
+  timestamp_sec:
+    _type: uint32_t
+    _comment: Time of last client update (in Seconds)
```
That is not clear to me. What is the reference time here?
How many seconds have passed relative to what time???

itayx replied:
@adam1985d correct me if i'm wrong
It's not relative to anything special, it's a timestamp stored as seconds
using `std::chrono::steady_clock::now()` or some other equivalent

adam1985d replied:
exactly it

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,98 @@ void son_management::handle_bml_message(Socket *sd,
+        }
+
+        auto response = message_com::create_vs_message<
+            beerocks_message::cACTION_BML_CLIENT_GET_CLIENT_LIST_RESPONSE>(cmdu_tx,
```
Why do you pass `beerocks_header->id()` everywhere? Only the controller tasks and who replying them should add it.

itayx replied:
Didn't know they were unneeded, I used the DCS APIs as a reference 
Removed

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,98 @@ void son_management::handle_bml_message(Socket *sd,
+        //To make sure the response is handled using a faux empty list.
+        std::string faux_client_list = "";
+        if (!response->alloc_client_list(faux_client_list.size())) {
+            LOG(ERROR) << "Failed client_list allocation to size=" << int(faux_client_list.size());
```
I don't think casting to int is needed for printing size_t.

itayx replied:
I'm not sure myself as well, but other places do cast `size_t` to `int` when printing so I did the same.
Changed anyway

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,98 @@ void son_management::handle_bml_message(Socket *sd,
+        }
+
+        //To make sure the response is handled using a faux empty list.
+        std::string faux_client_list = "";
```
I don't understand the `faux` thing.
If you plan to fill the list with real data in the future, just say it and the comment and explain that currently, you return an empty list.
Please don't write it as `faux`.

itayx replied:
Removed the mention of the Adjective faux, which basically means "fake"
And updated the comment to better reflect it's purpose

morantr commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,98 @@ void son_management::handle_bml_message(Socket *sd,
+            break;
+        }
+
+        std::copy_n(request->sta_mac().oct, MAC_ADDR_LEN, response->client().sta_mac.oct);
```
why not do this?
```cpp
response->client().sta_mac = request->sta_mac();
```

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,50 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * Get client list.
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [in/out] client_list      Client list.
```
On Doxygen, you need to use [in,out] and not [in/out].

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,50 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * Get client list.
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [in/out] client_list      Client list.
```
On Doxygen, you should put only one space between the argument name and its description.
I agree that it looks much better like you did it, but these are the rules. :-)

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,50 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * Get client list.
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [in/out] client_list      Client list.
```
I don't understand the client list format. Is it a string of MAC addresses separated by commas? Please add an explanation to the argument description.

itayx replied:
Expended the description

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,50 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * Get client info.
+ *
+ * @param [in] ctx                  BML Context.
+ * @param [in] sta_mac              MAC address of station.
```
"... of **a** station".
Need to add it also on `bml_client_dump_client` description..

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,50 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+ * @param [in/out] client_list      Client list.
+ * @param [in/out] client_list_size Size of client list.
+ *
+ * @return BML_RET_OK on success.
```
Why do you put an empty line between the @param and the @return? The empty line should be only after the @brief.

itayx replied:
Removed the unneeded empty line

morantr commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,60 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
+        mapf::utils::copy_string((char *)(client_list), temp_client_list.c_str(),
+                                 *client_list_size);
+        return BML_RET_OK;
+    }
```
It would look much nicer if you put the function on the `ret` declaration and not on the if statement.
Also please use `reinterpret_cast<char *>`. But why do you cast it? It is already `char *`
```cpp
    int ret = BML->client_get_client_list(temp_client_list, client_list_size);
    if (!ret) {
        mapf::utils::copy_string(client_list, temp_client_list.c_str(),
                                 *client_list_size);
        return BML_RET_OK;
    }
```

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,60 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
+
+    // Clear the outgoing arguments
+    *client_list_size = 0;
+    *client_list      = 0;
```
This is a "get" function, so why do you "delete" the things you got?
Also if it is a string you should put '\0' and not 0.

itayx replied:
This is a failed scenario, the "returning" values are incorrect
But changed the 0 to '\0' for the `client_list`

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2144,6 +2197,89 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+ */
+int cli_bml::client_get_client_list()
+{
+    std::cout << __func__ << std::endl;
```
I guess you added all of these `cout << __func__` print for debug, because these prints are not something we would like to see on the cli output. Please remove them all.

kantera800 commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    // 1 for true, 0 for false, -1 for "parameter is not configured".
+    int support_steering;
+    // 1 for true, 0 for false, -1 for "parameter is not configured".
+    int dual_band;
```
since there could be 3 bands shouldn't it be appropriate to have instead just single_band ?

itayx replied:
@adam1985d WDYT?

adam1985d replied:
I am actually not sure what the actual use of this parameter is.
it is mentioned specifically in the requirements (to be provided in 2nd phase) but perhaps as 6G is introduced, its meaning can change and it will become obsolete.
@yaniv-yitzhak WDYT?

itayx replied:
Changed to `single_band`

kantera800 commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    // 1 for true, 0 for false, -1 for "parameter is not configured".
+    int dual_band;
+    // Optional parameter
+    // The timelife delay of the client configuration (if configured), otherwise 0.
```
you should mention what zero stands for , not sure we even want to allow the user to set lifetime zero since then there is not point adding the configuration to the DB

itayx replied:
@adam1985d WDYT?

adam1985d replied:
if we eventually allow the a per-user configuration of time-life-delay, see [comment](https://github.com/prplfoundation/prplMesh/pull/1479/files#r448840254) then 0 means the parameter is not configured. you can also use it as `int` and provide -1 as `not-configured`,
but @kantera800 is right here, you need to define the meaning.

lets see what the comment mentioned above result in - we might remove it completely

kantera800 replied:
-1 as  `not-configured` seems good enough ,  just update the comment

itayx replied:
Updated

kantera800 commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    // 1 for true, 0 for false, -1 for "parameter is not configured".
+    int support_11v;
+    // 1 for true, 0 for false, -1 for "parameter is not configured".
+    int support_steering;
```
These parameters should be only read only for the user. 
IMO should represent  responsiveness [%]

```suggestion
    // client 11k responsiveness ratio in % since beginning of lifetime configuration".
    int support_11k_responsiveness;
    // client 11v responsiveness ratio in % since beginning of lifetime configuration".
    int support_11v_responsiveness;
    // client steering responsiveness ratio in % since beginning of lifetime configuration".
    int support_steering_responsiveness;
```

yaniv-yitzhak replied:
Why we need all those parameters here?

kantera800 replied:
@yaniv-yitzhak since we don't have to decide (for now ) on client's support to 11k/11v/steering  we could at least give the user the ability to  get client responsiveness (accumulated since first connection) statistics

itayx replied:
@adam1985d WDYT?

adam1985d replied:
these parameters were mentioned when the client's behavior was described in the requirements.
if they are not needed (or they should be provided in some other form like responsiveness) - we can remove/change accordingly - but thought this client **info** would be beneficial for the user

itayx replied:
These parameters are currently not used and are unneeded as these APIs are only the infrastructure phase, when we decide on the actual values we can expand as needed.
So for now these are removed

kantera800 commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,52 @@ sBssidInfo:
+    _comment: 1 for true, 0 for false, -1 for "not configured".
+  dual_band:
+    _type: int8_t
+    _comment: 1 for true, 0 for false, -1 for "not configured".
```
same here see my comment on the BML_CLIENT structure

itayx replied:
Fixed

kantera800 commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -686,7 +686,7 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
                              bool is_single_scan)
 {
-    // Validate intput params;
+    // Validate input parameters
```
please split all non related stuff  to a separate commit

itayx replied:
Done

kantera800 commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -557,6 +557,14 @@ void cli_bml::setFunctionsMapAndArray()
         INT_ARG);
+    insertCommandToMap("bml_client_get_client_list", "", "Get client list.",
+                       static_cast<pFunction>(&cli_bml::client_get_client_list_caller), 0, 0);
+    insertCommandToMap("bml_client_set_client", "<sta_mac>", "Set client with the given STA MAC.",
```
if this API only adds client mac to the DB with default values then where are the rest of APIs that configure rest of client's parameters ?

itayx replied:
Extended the CLI usage, and now the client needs to add the client parameters

kantera800 reviewed (changes requested):
In general look ok.
Although some of the client bml parameter description/naming is not accurate 
also,  set client parameters APIs are missing .

yaniv-yitzhak commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
+    char selected_bands[BML_SELECTED_BANDS_LEN];
+    // Optional parameter
+    // The timelife delay of the client configuration (if configured), otherwise 0.
+    uint32_t time_life_delay_days;
```
Remove this parameter, it should not be set per client, move it to conf file or UCI DB, this is not runtime configuration.

kantera800 replied:
if you want to give the client the ability to read it at least then it has to be here

yaniv-yitzhak replied:
No need, he can read it from conf file

adam1985d replied:
@yaniv-yitzhak
we said we will have the ability for per-client time-life delay configuration (of course it is optional and if not set the global parameter value will be used (according to friendly/unfriendly device))
if this is for sure not required I will also remove it from the controller-db, so need to be sure about it.

itayx replied:
I'll remove it from the BML_CLIENT_CONFIG, which is used in the "set" API
But keep it in the BML_CLIENT, which is used in the "get" API
Making it virtually as a "read only"

kantera800 replied:
@yaniv-yitzhak `time_life_delay_days`  parameter shouldn't  be part of conf file from these reasons:
1. we would need to handle it on boot.
2. no future ability to configure it per client.

either keep it in bml for read only or remove it completely and use define for aging.

adam1985d replied:
there are 2 layers here:
1. global params used by default (defined in prplmesh **config** file):
1.1. max lifetime delay - used for regular clients
1.2. unfriendly max lifetime delay - used for unfriendly clients
2. per-client (can be defined **per-client** in the **prplmesh-db**):
2.1 per-client lifetime delay - it is optional - if set it is used instead of the global parameters

- if we wish to support per-client configuration we must have both
- although we don't provide BML interface to configure the **global params** - user can edit them via UCI and restart prplmesh - so it is good to have them in **prplmesh config** file

kantera800 replied:
> 1. global params used by default (defined in prplmesh **config** file):
>    1.1. max lifetime delay - used for regular clients
>    1.2. unfriendly max lifetime delay - used for unfriendly clients

didn't see this mentioned in the HLD document

> * although we don't provide BML interface to configure the **global params** - user can edit them via UCI and restart prplmesh - so it is good to have them in **prplmesh config** file

it that case you need to add these parameters to aging handling mechanism on boot- not sure we have time for that.

I prefer to give access to less parameters as possible at least for now.

adam1985d replied:
I don't think there should be aging to these parameters - it is platform configuration just like any other prplmesh-config parameters

aging is only for clients' data/config - not for general platform configurations

yaniv-yitzhak commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +613,42 @@ struct BML_NEIGHBOR_AP {
 
+struct BML_CLIENT_CONFIG {
+    // 1 for true, 0 for false, -1 for "not configured".
+    int stay_on_initial_radio;
```
What's the difference between 0 & -1 -> both are actually false, why not using bool?

itayx replied:
There's a difference between "not configured" and false,
since false means it was configured and set to false.
Currently the notion of having "not configured" in the BML is to signify the client that that parameter is not persistent according to the DB

morantr replied:
> What's the difference between 0 & -1 -> both are actually false, why not using bool?

@yaniv-yitzhak, actually, '-1' is counts as "true".
'true' is defined as not "0".

itayx committed:
controller: bml: fix for comment

Small fix for two DCS comments

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

morantr commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +616,27 @@ struct BML_NEIGHBOR_AP {
+    int8_t stay_on_initial_radio;
+    // 1 for true, 0 for false, -1 for "not configured".
+    int8_t stay_on_selected_device;
+    // comma-separated selected bands [2.4,5H,5L,6].
```
Please start comments with a capital letter.
and it would look more clear if there will be a space between a parameter to the next parameter comment.
```cpp
// Comment on foo
int foo;

// Comment about bar;
int bar;
```

itayx replied:
Changed

morantr commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -610,6 +616,27 @@ struct BML_NEIGHBOR_AP {
+};
+
+struct BML_CLIENT : BML_CLIENT_CONFIG {
+    // Client MAC
```
I wouldn't write a comment about the obvious. (nitpick).

But please make sure to put a dot at the end of the sentence. Missing on the comment below too.

itayx replied:
Left the comment, without it it looks weird being the only parameter without a comment.
But added a dot at the end of the sentence.

morantr commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,60 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
+
+    // Clear the outgoing arguments
+    *client_list_size = 0;
+    *client_list      = '\0';
```
I commented on it last time.
You wrote:
>This is a failed scenario, the "returning" values are incorrect
But changed the 0 to '\0' for the client_list.

I didn't notice that inside the `if` block is a success scenario.
Usually, we handle errors, and then continue with the code. So in this case where 0 means success it should look like this:
```cpp
if (ret < 0) {
   // handle error
}
// rest of the code.
```

Also, in case of an error, you should not change the output arguments. The user should not use them either way.

itayx replied:
Changed to
```
    int ret = pBML->client_get_client_list(temp_client_list, client_list_size);
    if (!ret) {
        beerocks::string_utils::copy_string(client_list, temp_client_list.c_str(),
                                            *client_list_size);
    }

    return ret;
```

morantr commented on controller/src/beerocks/bml/bml.h:
```diff
@@ -648,6 +648,46 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
+int bml_client_get_client(BML_CTX ctx, const char *sta_mac, struct BML_CLIENT *client);
+
+/**
+ * Print client information
```
Missing dot at the end of the sentence.
Same on bml_internal.h.

itayx replied:
Fixed

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1307,6 +1321,52 @@ int cli_bml::get_dcs_scan_results_caller(int numOfArgs)
+ * caller function for client_get_client_list_caller
+ *
+ * @param [in] numOfArgs Num of received arguments
+ *
```
Please remove spaces between `@param` and `@return`.
Also missing dot at the end of the sentence.
Also (2), Start with a capital letter.
All are repeated several times on the file.

itayx replied:
Fixed

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1307,6 +1321,52 @@ int cli_bml::get_dcs_scan_results_caller(int numOfArgs)
+/**
+ * caller function for client_set_client_caller
+ *
+ * @param [in] numOfArgs Num of received arguments
```
Usually, Doxygen documentation should not have word shortcuts like "Num". Just write "Number".

itayx replied:
Fixed

morantr reviewed (approved):
Cosmetic only, so approving.

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -122,6 +122,12 @@ extern "C" {
 #define BML_CHANNEL_SCAN_ENUM_LIST_SIZE 8
 
+/* BML Client Steering */
+#define BML_CLIENT_SELECTEDBANDS_DISABLED 0
```
is it really that beneficial to omit the `_` in `...SELECETED_BANDS...` - this makes it much less readable

I also don't understand why all are `0`

if you meant to provide an `enum-like` functionality it should be similar to how `/* BML Node States */` or `/* BML Wi-Fi Security Modes */` are defined

itayx replied:
Resolved in https://github.com/prplfoundation/prplMesh/pull/1525

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -122,6 +122,12 @@ extern "C" {
+#define BML_CLIENT_SELECTEDBANDS_24G 0
+#define BML_CLIENT_SELECTEDBANDS_5G 0
+#define BML_CLIENT_SELECTEDBANDS_BOTH 0
+
```
I think the commit message is outdated - please update to what is actually implemented

and add `PPM-5` to all commit messages for auto-linking it to the JIRA issue

itayx replied:
Resolved in https://github.com/prplfoundation/prplMesh/pull/1525

adam1985d commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h:
```diff
@@ -415,7 +415,15 @@ enum eActionOp_BML: uint8_t {
+    ACTION_BML_CLIENT_GET_CLIENT_RESPONSE = 0xdc,
+    ACTION_BML_CLIENT_DUMP_CLIENT_REQUEST = 0xdd,
+    ACTION_BML_CLIENT_DUMP_CLIENT_RESPONSE = 0xde,
+    ACTION_BML_ENUM_END = 0xe1,
```
commit message suggestion:
replace
`Add the following message to be used later by the steering APIs.`
with
`Added client set/get/dump CMDUs to enable smart-steering:`

itayx replied:
Resolved in https://github.com/prplfoundation/prplMesh/pull/1525

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,47 @@ sBssidInfo:
+    _comment: 1 for true, 0 for false, -1 for "not configured".
+  time_life_delay_days:
+    _type: uint32_t
+    _comment: optional parameter, Timelife delay of the client configuration (if configured), otherwise 0.
```
perhaps we wish the `time_life_delay_days` to be of type int - to allow negative value for "not-configured" and "0" can have some other meaning - like "never-age"

itayx replied:
Changed & updated the comment

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,47 @@ sBssidInfo:
+  eSelectedBands_Disabled: 0
+  eSelectedBands_24G: 1
+  eSelectedBands_5G: 2
+  eSelectedBands_Both: 3
```
here you used different values than the zeros in the defines I commented in the previous commit. so I guess the previous commit had a bug

itayx replied:
Resolved in https://github.com/prplfoundation/prplMesh/pull/1525

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,94 @@ void son_management::handle_bml_message(Socket *sd,
+        }
+
+        request->sta_mac()                         = response->client().sta_mac;
+        response->client().timestamp_sec           = 0;
```
I know these are going to be changed once the controller DB PR is in, but in case this PR is inserted before - shouldn't the values be set as -1 (not-configured)?

itayx replied:
Changed the values that accept the `-1 (not-configured)` to return that until we integrate the DB

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1767,6 +1839,176 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+
+int bml_internal::client_dump_client(const sMacAddr &sta_mac)
+{
+    LOG(DEBUG) << "client_dump_client";
```
you are missing the implementation of the `client_dump_client`

itayx replied:
Removed the `client_dump_client` since the `client_get_client` prints the selected client in the BML CLI. so I don't see the need for a `dump_client` flow.

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1307,6 +1321,49 @@ int cli_bml::get_dcs_scan_results_caller(int numOfArgs)
+ */
+int cli_bml::client_get_client_list_caller(int numOfArgs)
+{
+    if (numOfArgs == 0) {
```
we might need to think of an interface that provides "max-number-of-client-to-get" or "returned-clients-buffer-max-size" similar to the get-scan-results in the DCS.
@kantera800 what do you think?

itayx replied:
TBH, the only reason we need the result count in the DCS is because of the outgoing API we aligned the BML with the client demand. a "max-number-of-client-to-get" is not needed because we would either get the client list as a vector of strings (or any other container that fits) or we concatenate the strings into a single one (like in the current code).
both cases there is no need for a size variable, since there shouldn't be a scenario where we want to get N listings when there is M available (and N < M), so we should always try and get the most listings possible (no artificial max value)

kantera800 replied:
we can add it later on if needed

adam1985d commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -686,7 +686,7 @@ int bml_get_dcs_scan_results(BML_CTX ctx, const char *radio_mac,
                              bool is_single_scan)
 {
-    // Validate intput params;
+    // Validate input parameters
```
usually commits that changes existing code should come first (will also make it easier for you to rebase on PR change requests)
nitpick

itayx replied:
Moved

adam1985d reviewed (changes requested):
please address the comments - I wish to re-review afterward so setting as "request-changes"

kantera800 reviewed (approved):
some minor unresolved issues are still open but besides that you can merged it.

adam1985d commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -122,6 +122,15 @@ extern "C" {
+#define BML_CLIENT_SELECTED_BANDS_24 1
+#define BML_CLIENT_SELECTED_BANDS_5G 2
+#define BML_CLIENT_SELECTED_BANDS_6G 4
+#define BML_CLIENT_SELECTED_BANDS_60 8
```
```suggestion
#define BML_CLIENT_SELECTED_BANDS_24G 1
#define BML_CLIENT_SELECTED_BANDS_5G 2
#define BML_CLIENT_SELECTED_BANDS_6G 4
#define BML_CLIENT_SELECTED_BANDS_60G 8
```

adam1985d replied:
either use `G` on all or not at all

itayx replied:
Fixed

adam1985d commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h:
```diff
@@ -415,7 +415,13 @@ enum eActionOp_BML: uint8_t {
+    ACTION_BML_CLIENT_SET_CLIENT_REQUEST = 0xd9,
+    ACTION_BML_CLIENT_SET_CLIENT_RESPONSE = 0xda,
+    ACTION_BML_CLIENT_GET_CLIENT_REQUEST = 0xdb,
+    ACTION_BML_CLIENT_GET_CLIENT_RESPONSE = 0xdc,
```
if you decided to remove the `dump` CMDU - edit the comment message accordingly

itayx replied:
Removed

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,55 @@ sBssidInfo:
+    _comment: 1 for true, 0 for false, -1 for "not configured".
+  selected_bands:
+    _type: int8_t
+    _comment: Bitset of selected bands supported by the client
```
probably better to mention the `eClientSelectedBands` here
(nitpick)

itayx replied:
Updated

adam1985d commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -989,3 +989,55 @@ sBssidInfo:
+    _comment: 1 for true, 0 for false, -1 for "not configured".
+  selected_bands:
+    _type: int8_t
+    _comment: Bitset of selected bands supported by the client
```
same as above:
probably best to mention `eClientSelectedBands` here
(nitpick)

itayx replied:
Updated

adam1985d commented on controller/src/beerocks/bml/bml.cpp:
```diff
@@ -709,3 +709,44 @@ int bml_start_dcs_single_scan(BML_CTX ctx, const char *radio_mac, int dwell_time
     return pBML->start_dcs_single_scan(tlvf::mac_from_string(std::string(radio_mac)), dwell_time,
```
better commit message would be:
```
This PR enables the smart-steering feature which allows a user to configure a client's steering behavior. The configuration is persistent across prplmesh restarts and reboots.

Added BML API to get list of configured clients.
Added BML client set and get APIs.
added BML-internal implementation
```

itayx replied:
Resolved in https://github.com/prplfoundation/prplMesh/pull/1525

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1767,6 +1839,168 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+    if (!m_sockMaster) {
+        int iRet = connect_to_master();
+        if (iRet != BML_RET_OK) {
+            return iRet;
```
add error log print

itayx replied:
Added error log

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1767,6 +1839,168 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+        return iRet;
+    }
+
+    return BML_RET_OK;
```
```suggestion
    LOG_IF(iRet != BML_RET_OK, ERROR) << "Client list returned with error code:" << iRet;

    return iRet;
```

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1767,6 +1839,168 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+    if (!m_sockMaster) {
+        int iRet = connect_to_master();
+        if (iRet != BML_RET_OK) {
+            return iRet;
```
add error log print

itayx replied:
Added error log

adam1985d commented on controller/src/beerocks/bml/internal/bml_internal.cpp:
```diff
@@ -1767,6 +1839,168 @@ int bml_internal::start_dcs_single_scan(const sMacAddr &mac, int dwell_time_ms,
+        return iRet;
+    }
+
+    return BML_RET_OK;
```
```suggestion
    LOG_IF(iRet != BML_RET_OK, ERROR) << "Client returned with error code:" << iRet;
    
    return iRet;
```

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,90 @@ void son_management::handle_bml_message(Socket *sd,
         LOG(TRACE) << "ACTION_BML_CHANNEL_SCAN_DUMP_RESULTS_REQUEST";
         break;
     }
```
an alternative commit message:

Added CMDU-handling stubs for smart-steering BML messages in son_management.

adam1985d commented on controller/src/beerocks/master/son_management.cpp:
```diff
@@ -2308,6 +2308,90 @@ void son_management::handle_bml_message(Socket *sd,
+
+        std::string client_list = "";
+        // client_list will be filled with client information retrived later on.
+        // Currently an empty string is returned for testing purposes.
```
```suggestion
        // TODO: replace empty string with a list of configured clients read from controller-db
```

itayx replied:
Changed ToDO comment

adam1985d commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -474,6 +474,8 @@ enum class eChannelScanOperationCode : uint8_t {
 #define CHANNEL_SCAN_INVALID_PARAM -1
 #define SCAN_ALL_CHANNELS 0
 
+#define CLIENT_PARAMETER_NOT_CONFIGURED -1
```
```suggestion
#define PARAMETER_NOT_CONFIGURED -1
```

why limit to just client parameters - this define can be used to any parameter

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -557,6 +557,20 @@ void cli_bml::setFunctionsMapAndArray()
+        " selected_bands - 1 for 2.4G, 2 for 5G, 3 for both, 0 for Disabled"
+        " stay_on_initial_radio - 1 for true, 0 for false"
+        " stay_on_selected_device - 1 for true, 0 for false",
+        static_cast<pFunction>(&cli_bml::client_set_client_caller), 4, 4, STRING_ARG, INT_ARG,
```
I think we can support 2->4 parameters where the mac is always the first and the rest need to be provided in an "optional-params" syntex.
cfr. `bml_set_dcs_continuous_scan_params` implementation

itayx replied:
Changed

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1307,6 +1321,49 @@ int cli_bml::get_dcs_scan_results_caller(int numOfArgs)
+int cli_bml::client_set_client_caller(int numOfArgs)
+{
+    if (numOfArgs == 4) {
+        return client_set_client(args.stringArgs[0], args.intArgs[1], args.intArgs[2] == 1,
```
originally I liked how you immediately changed to bool where you could!

but since these can be optional (as commented above) this implementation of `client_set_client` should change to support optional-cli-function-params.

itayx replied:
Changed to support the optional parameters (as not-configured)

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2144,6 +2201,104 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+ * @param [in] stay_on_selected_device Whather to stay on selected device or not.
+ * @return 0 on success.
+ */
+int cli_bml::client_set_client(const std::string &sta_mac, int8_t selected_bands,
```
see my comment above regarding optional parameters and how you enforce here a true/false even though the `BML_CLIENT_CONIG` supports optional params

itayx replied:
Changed to use optional parameters

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -557,6 +557,20 @@ void cli_bml::setFunctionsMapAndArray()
+        "bml_client_set_client",
+        "<sta_mac> <selected_bands> <stay_on_initial_radio> <stay_on_selected_device>",
+        "Set client with the given STA MAC:"
+        " selected_bands - 1 for 2.4G, 2 for 5G, 3 for both, 0 for Disabled"
```
the description here should be aligned to the new implementation using the bitwise

itayx replied:
Updated the description

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2144,6 +2201,104 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+                  << " stay_on_initial_radio: " << client_bool_print(client.stay_on_initial_radio)
+                  << std::endl
+                  << " stay_on_selected_device: "
+                  << client_selected_bands_print(client.selected_bands) << std::endl
```
you have an error here in the print
reorganize it

itayx replied:
Fixed it

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2144,6 +2201,104 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+                  << " stay_on_selected_device: "
+                  << client_selected_bands_print(client.selected_bands) << std::endl
+                  << client_bool_print(client.stay_on_selected_device) << std::endl
+                  << " selected_bands: " << client.selected_bands << std::endl
```
use the lambda function here?

itayx replied:
Fixed the print

adam1985d commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -214,6 +217,10 @@ class cli_bml : public cli {
                              bool is_single_scan = false);
+    int client_get_client_list();
+    int client_set_client(const std::string &sta_mac, int8_t selected_bands,
+                          bool stay_on_initial_radio, bool stay_on_selected_device);
```
the function should receive int8_t to support "not-configured", cfr. `set_dcs_continuous_scan_params` function

itayx replied:
Changed the function to support "not-configured" params

adam1985d reviewed (changes requested):
please fix comments and trigger re-review

also for all commits:
mention PPM-5 in the commit message for auto-link to the task in JIRA

I know this PR had several "rework" stages - make sure we are aligned to the up-to-date way of implementation in all layers (common/BML/BML-internal/son-management).

itayx commented:
This PR is replaced by https://github.com/prplfoundation/prplMesh/pull/1525
